server/src/main/java/org/elasticsearch/cluster/metadata/MetadataAutoshardIndexService.java

```diff
@@ -0,0 +1,171 @@
+                    delegate.map(shardsAcknowledged -> {
+                        if (shardsAcknowledged == false) {
+                            logger.debug(
+                                "[{}] index autoshard complete, but the operation timed out while waiting for enough shards to be started.",
+                                request.index()
+                            );
+                        } else {
+                            logger.trace("[{}] index autoshard complete and shards acknowledged", request.index());
+                        }
+                        return ShardsAcknowledgedResponse.of(true, shardsAcknowledged);
+                    })
+                );
+            } else {
+                logger.trace("index creation not acknowledged for [{}]", request);
+                delegate.onResponse(ShardsAcknowledgedResponse.NOT_ACKNOWLEDGED);
+            }
+        }));
+    }
+
+    private void onlyAutoshardIndex(
+        final TimeValue masterNodeTimeout,
+        final TimeValue ackTimeout,
+        final AutoshardIndexClusterStateUpdateRequest request,
+        final ActionListener<AcknowledgedResponse> listener
+    ) {
+        var delegate = new AllocationActionListener<>(listener, threadPool.getThreadContext());
+        submitUnbatchedTask(
+            "autoshard-index [" + request.index() + "], cause [" + request.cause() + "]",
+            new AckedClusterStateUpdateTask(Priority.URGENT, masterNodeTimeout, ackTimeout, delegate.clusterStateUpdate()) {
+
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    return applyAutoshardIndexRequest(currentState, request, false, delegate.reroute());
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    if (e instanceof ResourceAlreadyExistsException) {
+                        logger.trace(() -> "[" + request.index() + "] failed to create", e);
+                    } else {
+                        logger.debug(() -> "[" + request.index() + "] failed to create", e);
+                    }
+                    super.onFailure(e);
+                }
+            }
+        );
+    }
+
+    public ClusterState applyAutoshardIndexRequest(
+        ClusterState currentState,
+        AutoshardIndexClusterStateUpdateRequest request,
+        boolean silent,
+        final ActionListener<Void> rerouteListener
+    ) {
+        final String indexName = request.index();
+        final IndexMetadata sourceMetadata = currentState.metadata().index(indexName);
+        int sourceNumShards = sourceMetadata.getNumberOfShards();
+
+        Settings.Builder settingsBuilder = Settings.builder().put(sourceMetadata.getSettings());
+        settingsBuilder.remove(IndexMetadata.SETTING_NUMBER_OF_SHARDS);
+        settingsBuilder.put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, sourceNumShards * 2);
```

Comment:
I suppose it might be for later, but seems like it might be nice for the new number of shards to be a request parameter.